src/sqlfluff/rules/references/RF03.py

```diff
@@ -93,14 +92,24 @@ def _eval(self, context: RuleContext) -> EvalResultType:
         if context.dialect.name in self._dialects_with_structs:
             self._is_struct_dialect = True
 
-        if not FunctionalContext(context).parent_stack.any(sp.is_type(*_START_TYPES)):
+        # If any of the parents would have also triggered the rule, don't fire
+        # because they will more accurately process any internal references.
+        if not any(seg.is_type(*_START_TYPES) for seg in context.parent_stack):
```

Comment:
No need to convert to the functional context here, doing it with builtins is more readable and simpler.